src/Build/Evaluation/ItemSpec.cs

```diff
@@ -446,6 +461,12 @@ public GlobFragment(string textFragment, string projectDirectory)
         {
         }
 
+        public override IEnumerable<string> GetReferencedItems()
+        {
+            // This fragment cannot efficiently enumerate its referenced items.
+            return null;
```

Comment:
```suggestion
            return Enumerable.Empty<string>();
```

Reply:
This would make it hard to differentiate between "there were no referenced items" and "we couldn't enumerate them efficiently." Also, I don't think it's the critical case we're trying to optimize for here—I believe we're mostly looking at RemoveOperations without globs/wildcards.

Reply:
I actually spent some time trying to come up with a combination of glob Includes and Removes that would be slow but couldn't find anything bad. It is orders of magnitude faster than the scenarios addressed with the PR.

This `null` is important, it tells the caller that rather than enumerating all referenced items and removing them from the running list, the caller should check each item in the running list and test it against the glob via `IsMatch`.

Reply:
> it tells the caller that rather than enumerating all referenced items and removing them from the running list, the caller should check each item in the running list and test it against the glob via IsMatch

Yup, which makes the calling code more complex (see https://github.com/dotnet/msbuild/pull/6716#discussion_r678762503)